### PR TITLE
Parameter.set_to: use get_latest, not _latest['value']

### PR DESCRIPTION
**Pull request by astafan8**
Refactor `Paramter.set_to` context manager to NOT directly use `._latest` cache. The idea also is that `set_to` defers the parameter-has-not-even-been-captured-hence-we-first-need-to-call-get logic to `get_latest()` as opposed to implementing it itself. Moreover, this PR adds some extra tests for `set_to(None)` edge cases.

@QCoDeS/core when reviewing, please let me know if I missed a test case so that I can add it to the test suite and hence ensure that this refactoring does not break the uncaptured behavior (note that I already added one extra test)

### CODE REVIEW

File: qcodes/instrument/parameter.py [hunk 1]
```diff
@@ -115,15 +115,22 @@ class _SetParamContext:
...
+        )
 
-        if self._original_value is None and self._value_is_changing:
-            self._original_value = self._parameter.get()  # type: ignore[has-type]
+        if parameter_has_been_captured_before or self._value is not None:
+            self._original_value = self._parameter.get_latest()
+            self._value_is_changing = self._value != self._original_value
+        else:
+            self._original_value = None
+            self._value_is_changing = False
```
[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/instrument/parameter.py]: i think i half-understood the use-case but still not completely in order to write a test that represents it - could you (pseudo)code it here? (also, i don't understand what snapshot has to do with all of this)

does the following depict the case that you have in mind? :
```python
# representing instrument here
instr_value = 123
set_counter = 0

def set_instr_value(val):
    nonlocal instr_value, set_counter
    instr_value = value
    set_counter += 1

# make a parameter that is linked to an instrument
p = Parameter('p', set_cmd=set_instr_value, get_cmd=lambda: instr_value)

# pre-conditions
assert p._latest['value'] is None
assert p._latest['raw_value'] is None
assert p._latest['ts'] is None
assert set_counter == 0

with p.set_to(None):
    # Do you mean that this expectation is incorrect
    # because the user wanted the parameter to be set 
    # to `None` but the parameter `p`
    # doesn't know about the `123` because its `_latest` is
    # not in sync with the instrument?
    assert set_counter == 0  # is this a wrong expectation?
    assert instr_value == None  # is this a wrong expectation?
assert set_counter == 0
assert instr_value == 123
```

[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/instrument/parameter.py]: I think snapshot should read cache sorry

[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/instrument/parameter.py]: But I am more curious if you can explain why this if / else clause cannot simply be replaced by the if case. 
I cannot see when the else clause would be correct

[the remark this answers is not in the document]
Comment (Dominik-Vogel) [reply qcodes/instrument/parameter.py]: I have a much simpler doubt about the same thing:
The parameter has not been captuered, then we set it to something, but don't set it back to the unknown value.

[the remark this answers is not in the document]
Comment (Dominik-Vogel) [reply qcodes/instrument/parameter.py]: Can we not simply leave it out:
```python
class _SetParamContext:
    """
    This class is returned by the ``set_to`` method of parameters
    Example usage:
    >>> v = dac.voltage()
    >>> with dac.voltage.set_to(-1):
        ...     # Do stuff with the DAC output set to -1 V.
        ...
    >>> assert abs(dac.voltage() - v) <= tolerance
    """
    def __init__(self, parameter: "_BaseParameter", value: ParamDataType):
        self._parameter = parameter 
        self._temp_value = value
        self._original_value = self._parameter.get_latest()

    def _value_is_changing(self) -> bool:
        return self._value != self._original_value

    def __enter__(self) -> None:
        if self._value_is_changing():
            self._parameter.set(self._value)  # type: ignore[has-type]

    def __exit__(self, typ,  # type: ignore[no-untyped-def]
                 value,
                 traceback) -> None:
        if self._value_is_changing:
            self._parameter.set(self._original_value)  # type: ignore[has-type]
```

[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/instrument/parameter.py]: This is what will currently happen as I read your code

```
# representing instrument here
instr_value = 123
set_counter = 0

def set_instr_value(val):
    nonlocal instr_value, set_counter
    instr_value = value
    set_counter += 1

# make a parameter that is linked to an instrument
p = Parameter('p', set_cmd=set_instr_value, get_cmd=lambda: instr_value)

# pre-conditions
assert p._latest['value'] is None
assert p._latest['raw_value'] is None
assert p._latest['ts'] is None
assert set_counter == 0

with p.set_to(None):
    # in the context manager creation 
    # parameter_has_been_captured_before  = False
    # value is None  we therefore enter the second path (the else clause)

   # now _original_value is set to None which is wrong (since the value on the instrument is 123
   # and _value_is_changing  is False which is also wrong since the value on the instrument should change from 123 to whatever None represents.
    assert set_counter == 0 # No actual set is performed but it should have been
    assert instr_value == 123  # This will be 123 since no set has been performed
assert set_counter == 0 # set counter is still 0 sinse value_changing is false the value is never set back
assert instr_value == 123 # the value is unchanged after the context manager so that is fine

```

[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/instrument/parameter.py]: > why this if / else cannot simply be replaced by the if case

it cannot because then `self._original_value` and `self._value_is_changing` are undefined. What should they be in that case? Do you mean to suggest that `if not parameter_has_been_captured_before is True` then `self._value_is_changing is True` even if `value is None`? (read another suggestion below as well)

> The parameter has not been captured, then we set it to something, but don't set it back to the unknown value.

hmm, i think the issue is that we don't have a sentinel that means "parameter-has-not-been-captured" and `None` is sometimes used for it, but can also be a legit value of the parameter.

shall we raise in the current `else` clause saying smth like "parameter hasn't been captured before and the new value to set to is None, qcodes doesn't know what to do, please 'get' the parameter first and then try again to 'set_to'"?

[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/instrument/parameter.py]: @Dominik-Vogel sorry, missed your suggested code about that remove the if clause alltogether. I think i like it very much. let me try it and see which tests fail......

[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/instrument/parameter.py]: @jenshnielsen i'll add this test taking converting assertions from "wrong" to correct.

[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/instrument/parameter.py]: > it cannot because then self._original_value and self._value_is_changing are undefined. What should they be in that case? > Do you mean to suggest that if not parameter_has_been_captured_before is True then self._value_is_changing is True 
> even if value is None? (read another suggestion below as well)

This statement is simply wrong as far as I can see. If `ts is None` `get_latest()` will call `get()` so how can it be undefined

[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/instrument/parameter.py]: @Dominik-Vogel so the implementation that you suggested does not pass [the new test test_set_to_none_when_current_value_is_none that I added](https://github.com/QCoDeS/Qcodes/pull/1786/files#diff-fe69d76e27bbbfbafc041a23a1003a32R1606). May I ask you to have a look at that test and let me know if the test itself makes sense?

[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/instrument/parameter.py]: @jenshnielsen my initial implementation does NOT pass "this new test". @Dominik-Vogel 's implementation does pass "this new test". Pushing "this new test" now ....

[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/instrument/parameter.py]: Ok that is fine. There is no difference between what I suggest and what @Dominik-Vogel suggests The only difference is moving the is changing call from construction time to enter time which is fully equivalent

File: qcodes/tests/test_parameter.py [hunk 2]
```diff
@@ -1594,14 +1595,47 @@ def _pp_setter(self, value):
...
+        assert self._cp_counter == 0
+        assert self._cp_get_counter == 0
+        assert counting_parameter._latest["value"] is None
+        assert counting_parameter.get_latest.get_timestamp() is None
+
+        with counting_parameter.set_to(None):
+            # The value should not change
+            assert counting_parameter._latest["value"] is None
+            # The timestamp of the latest value should be still None
+            assert counting_parameter.get_latest.get_timestamp() is None
```
Comment (Dominik-Vogel) [inline qcodes/tests/test_parameter.py]: I think the test may be wrong.
To me the name suggests a different test. It says that the current value is none, where the test sets up timestamp to be none, which means that the current state is actually unknown.
If it is unknown we have to get it on entering the context, so that we can restore it at the end.
Therefore we have to expect the timestamp to change. (See marked line here)

File: qcodes/tests/test_parameter.py [hunk 3]
```diff
@@ -1594,14 +1595,47 @@ def _pp_setter(self, value):
...
+            # Set method is not called
+            assert self._cp_counter == 0
+            # Get method is not called
+            assert self._cp_get_counter == 0
```
Comment (Dominik-Vogel) [inline qcodes/tests/test_parameter.py]: (has to change as well)

File: qcodes/tests/test_parameter.py [hunk 4]
```diff
@@ -1594,14 +1595,47 @@ def _pp_setter(self, value):
...
+
+        # The value should not change
+        assert counting_parameter._latest["value"] is None
+        # The timestamp of the latest value should be still None
+        assert counting_parameter.get_latest.get_timestamp() is None
```
Comment (Dominik-Vogel) [inline qcodes/tests/test_parameter.py]: ...

File: qcodes/tests/test_parameter.py (hunk 2 above)
Comment (astafan8) [reply qcodes/tests/test_parameter.py]: got it. let me adjust the test....

File: qcodes/instrument/parameter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/instrument/parameter.py]: @jenshnielsen thanks! now also with Dominik's comments on the test I see what to do. Basically `set_to` should not worry about the `new_value` being `None` or not, it should COMPLETELY deffer that logic to `get_latest` (as was originally planned). I will push stuff shortly

File: qcodes/tests/test_parameter.py [hunk 5]
```diff
@@ -1594,14 +1595,82 @@ def _pp_setter(self, value):
...
+        # Pre-conditions:
+        assert self._cp_counter == 0
+        assert self._cp_get_counter == 0
+        assert counting_parameter._latest["value"] is None
+        assert counting_parameter.get_latest.get_timestamp() is None
+
+        with counting_parameter.set_to(None):
+            # The value should not change
+            assert counting_parameter._latest["value"] is None
+            # The timestamp of the latest value should be still None
```
Comment (jenshnielsen) [inline qcodes/tests/test_parameter.py]: This comment is out of date now

File: qcodes/tests/test_parameter.py [hunk 6]
```diff
@@ -1594,14 +1595,82 @@ def _pp_setter(self, value):
...
+            assert counting_parameter.get_latest.get_timestamp() is not None
+            # Set method is not called
+            assert self._cp_counter == 0
+            # Get method is called once
+            assert self._cp_get_counter == 1
+
+        # The value should not change
+        assert counting_parameter._latest["value"] is None
+        # The timestamp of the latest value should be still None
```
Comment (jenshnielsen) [inline qcodes/tests/test_parameter.py]: same

**Comment by astafan8** [comment]
@jenshnielsen fixed them.

**Comment by jenshnielsen** [comment]
Good to merge once CI passes
